sha3.py

```diff
@@ -0,0 +1,223 @@
+    0x8000000000008003,
+    0x8000000000008002,
+    0x8000000000000080,
+    0x000000000000800A,
+    0x800000008000000A,
+    0x8000000080008081,
+    0x8000000000008080,
+    0x0000000080000001,
+    0x8000000080008008,
+]
+
+
+def absorb_block(state: Tensor, block: bytearray) -> Tensor:
+  for idx in range(0, rate_bytes, 8):
+    block_val = block[idx: idx + 8]
+    uint64 = int.from_bytes(block_val, "little")
+    state[idx // 8] = state[idx // 8] ^ uint64
+  return state.view(height, width)
+
+
+def rotl64(n: Union[int, Tensor], shifts: int):
+  lshift = (n << shifts) & ((1 << 64) - 1)
+  rshift = n >> (64 - shifts)
+  return lshift | rshift
+
+
+def theta(state: Tensor) -> Tensor:
+  C = state[0] ^ state[1] ^ state[2] ^ state[3] ^ state[4]
+  left = C.roll(shifts=-4, dims=0)
+  right = C.roll(shifts=-1, dims=0)
+  rotated = rotl64(right, 1)
+  D = left ^ rotated
+  return state ^ D
+
+
+def rho_pi(state: Tensor) -> Tensor:
+  """
+  Combined rho and pi steps
+  """
+  rho_pi_state = Tensor.zeros(
+      height * width, dtype=dtypes.uint64).contiguous()
+  for i in range(height * width):
+    x, y = divmod(i, width)
+    rotated = rotl64(state[x][y], rho_offsets[x][y])
+    pos = pi_offsets[x][y]
+    rho_pi_state[pos] = rotated
+  return rho_pi_state.view(height, width)
+
+
+def chi(state: Tensor) -> Tensor:
+  shift1 = state.roll(shifts=-1, dims=1)
+  shift2 = state.roll(shifts=-2, dims=1)
+  # ~shift1
+  return state ^ ((shift1 ^ Tensor.full_like(shift1, -1)) & shift2)
+
+
+def iota(state: Tensor, round_idx: int) -> Tensor:
+  state[0][0] ^= iota_round_constants[round_idx]
+  return state
+
+
+def keccak_round(state: Tensor, round_idx: int) -> Tensor:
+  """
+  One round of Keccak-f
+  """
+  state = theta(state)
+  rho_pi_state = rho_pi(state)
+  chi_state = chi(rho_pi_state)
+  return iota(chi_state, round_idx)
+
+
+def keccak_f(state: Tensor) -> Tensor:
+  """
+  Keccack-f function
+  """
+  for round_idx in range(24):
+    state = keccak_round(state, round_idx)
+  return state
+
+
+def pad(state: Tensor, pad_idx_bytes: int):
+  # adjust padding offsets to get the correct byte inside the uint64
+  padpoint_uint = (pad_idx_bytes % rate_bytes) // 8
+  padpoint_shift = ((pad_idx_bytes % rate_bytes) % 8) * 8
+  state[padpoint_uint] = state[padpoint_uint] ^ (0x06 << padpoint_shift)
+  state[(rate_bytes // 8) - 1] = state[(rate_bytes // 8) - 1] ^ (0x80 << 56)
+  return state
+
+
+def visualize_state(state: Tensor):
+  """
+  Print the state for debugging
+  """
+  for x in range(height):
+    for y in range(width):
+      print(f"{state.view(height, width)[x][y].numpy():016x}", end=" ")
+    print()
+
+
+def digest(state: Tensor) -> str:
+  return state.flatten().numpy().tobytes()[:md_len].hex()
+
+
+def tinygrad_sha3_256(message: str) -> str:
+  msg_bytes = bytearray(message.encode("utf-8"))
+  state = Tensor.zeros(height * width, dtype=dtypes.uint64).contiguous()
+
+  # absorb
+  for block_idx in range(0, len(msg_bytes), rate_bytes):
```

Comment:
This is not parallel

Reply:
Updated to implement batch-level parallelism.